### PR TITLE
Distinguish errors raised during validation in get_or_create

### PR DESCRIPTION
**Pull request by ipanova**
[noissue]

### CODE REVIEW

File: pulpcore/app/serializers/base.py [hunk 1]
```diff
@@ -369,7 +369,19 @@ def get_or_create(cls, natural_key, default_values=None):
             try:
                 serializer.is_valid(raise_exception=True)
                 result = serializer.create(serializer.validated_data)
```
Comment (mdellweg) [inline pulpcore/app/serializers/base.py:371]: I think you can map these exception to these two lines individually.

**Review by mdellweg — APPROVED** [review]

File: pulpcore/app/serializers/base.py [hunk 2]
```diff
@@ -368,8 +368,21 @@ def get_or_create(cls, natural_key, default_values=None):
...
+                if error_codes.intersection(all_codes):
+                    # validation failed because field is invalid, missing
+                    raise e
+                # recover from a race condition, where another thread just created the object
+                # validation failed with 400 'unique' error code
+                result = cls.Meta.model.objects.get(**natural_key)
+            try:
                 result = serializer.create(serializer.validated_data)
-            except (IntegrityError, serializers.ValidationError):
+            except IntegrityError:
```
Comment (lubosmj) [inline pulpcore/app/serializers/base.py:385]: It looks like we create the object after fetching from the database? Lines 382-384

Comment (ipanova) [reply pulpcore/app/serializers/base.py:385]: maybe i will just remove L382 to no have have an extra DB call

File: pulpcore/app/serializers/base.py [hunk 3]
```diff
@@ -368,8 +369,21 @@ def get_or_create(cls, natural_key, default_values=None):
...
+                for c in returned_codes:
+                    all_codes.extend(c)
+                if error_codes.intersection(all_codes):
+                    # validation failed because field is invalid, missing
+                    raise e
+                # recover from a race condition, where another thread just created the object
+                # validation failed with 400 'unique' error code
+                result = cls.Meta.model.objects.get(**natural_key)
+            try:
+                result = result if result else serializer.create(serializer.validated_data)
```
Comment (mdellweg) [inline pulpcore/app/serializers/base.py:385]:
```suggestion
                result = result or serializer.create(serializer.validated_data)
```

File: pulpcore/app/serializers/base.py [hunk 4]
```diff
@@ -368,8 +369,21 @@ def get_or_create(cls, natural_key, default_values=None):
...
+                returned_codes = e.get_codes().values()
+                error_codes = set(["required", "invalid"])
+                all_codes = []
+                for c in returned_codes:
+                    all_codes.extend(c)
+                if error_codes.intersection(all_codes):
+                    # validation failed because field is invalid, missing
+                    raise e
+                # recover from a race condition, where another thread just created the object
+                # validation failed with 400 'unique' error code
```
Comment (mdellweg) [inline pulpcore/app/serializers/base.py:382]: I wonder if we could check for "unique" instead of "all but unique"?

Comment (ipanova) [reply pulpcore/app/serializers/base.py:382]: I need to raise error if at least one of error_codes are present. There can be a mixture of unique, required, invalid.

**Review by lubosmj — APPROVED** [review]

**Pull request merged**